ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -134,74 +120,72 @@ where
         <λ as Sub<U1>>::Output: ArrayLength,
-        I: IntoIterator<Item = F>,
-        J: IntoIterator<Item = F>,
-        I::IntoIter: ExactSizeIterator,
-        J::IntoIter: ExactSizeIterator,
+        J: Iterator<Item = B>,
+        B: Borrow<(GenericArray<F, λ>, GenericArray<F, λ>)>,
     {
-        let mut u = u.into_iter();
-        let mut v = v.into_iter();
-
-        debug_assert_eq!(u.len() % λ::USIZE, 0); // We should pad with zeroes eventually
-
-        let s = u.len() / λ::USIZE;
-
-        assert!(
-            s > 1,
-            "When the output is this small, you should validate the proof with a more straightforward reveal"
-        );
-
         let r: F = hash_to_field(
             &compute_hash(proof_left),
             &compute_hash(proof_right),
             λ::U128,
         );
-        let mut p = GenericArray::<F, λ>::generate(|_| F::ZERO);
-        let mut q = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        let mut output = Vec::<(GenericArray<F, λ>, GenericArray<F, λ>)>::new();
         let denominator = CanonicalLagrangeDenominator::<F, λ>::new();
         let lagrange_table_r = LagrangeTable::<F, λ, U1>::new(&denominator, &r);
 
-        let pairs = (0..s).map(|_| {
-            for i in 0..λ::USIZE {
-                let x = u.next().unwrap_or(F::ZERO);
-                let y = v.next().unwrap_or(F::ZERO);
-                p[i] = x;
-                q[i] = y;
+        // iter and interpolate at x coordinate r
+        let mut index = 0;
+        let mut new_u_chunk = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        let mut new_v_chunk = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        for polynomial in uv_iterator {
+            let (u_chunk, v_chunk) = polynomial.borrow();
+            let u = lagrange_table_r.eval(u_chunk)[0];
+            let v = lagrange_table_r.eval(v_chunk)[0];
+            if index >= λ::USIZE {
+                output.push((new_u_chunk, new_v_chunk));
+                new_u_chunk = GenericArray::<F, λ>::generate(|_| F::ZERO);
+                new_v_chunk = GenericArray::<F, λ>::generate(|_| F::ZERO);
+                index = 0;
             }
-            let p_r = lagrange_table_r.eval(&p)[0];
-            let q_r = lagrange_table_r.eval(&q)[0];
-            (p_r, q_r)
-        });
-        let (u, v) = pairs.unzip();
-        ProofGenerator::new(u, v)
+            new_u_chunk[index] = u;
+            new_v_chunk[index] = v;
+            index += 1;
+        }
+        if index != 0 {
+            output.push((new_u_chunk, new_v_chunk));
+        }
+
+        Self { uv: output }
     }
 }
 
-impl<F> PartialEq<(&[u128], &[u128])> for ProofGenerator<F>
+#[allow(non_camel_case_types)]
+impl<F, λ> PartialEq<(&[u128], &[u128])> for ProofGenerator<F, λ>
 where
     F: PrimeField + std::cmp::PartialEq<u128>,
+    λ: ArrayLength,
 {
     fn eq(&self, other: &(&[u128], &[u128])) -> bool {
         let (cmp_a, cmp_b) = other;
```

Comment:
I know I chose these names, but I think we can improve them. cmp_a and cmp_b don't tell me what they are.

Reply:
I think we currently only need this function for tests, so it would make sense to move it into the test section. I am happy to change the names despite this though.